Make tests on boolean build variables more readable

shindere opened the pull request:
We have several boolean configuration variables to specify whether
a feature is enabled or not. `NATIVE_COMPILER` is an example of such a
variable.

In our build system, we have several instances of tests like this:

```
ifneq "$(NATIVE_COMPILER)" "false"
```

Since it is so easy to misread such tests and to overlook the `n`,
the only purpose of this PR is to replace such tests by their more
readable equivalent:

```
ifeq "$(NATIVE_COMPILER)" "true"
```

Snce not all the poorly written tests were on an explicit boolean
variable, it was necessary to introduce a one, `lib_runtime_events` and
to make a few others more widely available within the build system, namely
the boolean variables introduced to let `ocamltest` know whether
the libraries and ocamldebug/ocamldoc are there.

One thing could still be improved: the current definition of `with_camltex`
is odd: it's true when the tool needs to be built but the empty string
(rather than false) when it is not necessary to build the tool.

The rewriting described above still works so the definition can be left
untouched. We could also fix it, but since `with_camltex` is defined in
`Makefile.config` there is a risk of breaking external tools that assume
the variable is the empty string when the tool is not available.
In looks unlikely to me that anyone even uses cmaltex outside of the
compiler's codebase, but I may be wrong on this. @Octachron?

One other possibility would be to introduce a second, sibbling
variable, `build_camltex` which would be defined correctly and which would
be used internally, while `with_camltex` would be left untouched.

shindere committed:
Introduce and use the lib_runtime_events boolean configuration variable

The configure script sets this variable to "true" if the runtime_events
library has to be built. At the moment it is always built but this may
change.

shindere committed:
Make more config vars visible to the whole build system and use them

shindere committed:
Rename configuration variables and make them more widely available

This commit renames the ocamltest_ocamldebug and ocamltest_ocamldoc
configuration variables to build_ocamldebug and build_ocamldoc, makes them
available to the whole build system and uses them to simplify a few
conditionals in the build system.

shindere committed:
build system: make a few conditionals over boolean variables more readable

Basically, this commit replaces make conditionals like

ifneq "$(NATIVE_COMPILER)" "false"

by the hopefully more readable

ifeq "$(NATIVE_COMPILER)" "true"

Octachron commented:
I don't think anyone rely on the specific format of the `with_camltex` variable in `Makefile.config`. The `caml-tex` tool is never installed and the variable only matters when building the manual.

shindere commented:
Florian Angeletti (2022/09/11 12:57 -0700):
> I don't think anyone rely on the specific format of the `with_camltex`
> variable in `Makefile.config`. The `caml-tex` tool is never installed
> and the variable only matters when building the manual.

OK, many thanks for the prompt response!

Do you mind if I then move the definition of the variable to
`Makefile.build_config.in` to make it clear it's private?

I also would like to rename camltex to ocamltex, for consistency
reasons and even if the tool is currently not installed. And of course
the name of the variable would then change to `build_ocamltex` and its
definition when the tool is disabled would change as described.

Would all this be okay?

Octachron commented:
Moving the variable to `Makefile.build_config.in` is definitively a good idea.
Renaming `caml-tex` to `ocamltex` seems fine since it fits with the various *tex tools.

shindere commented:
Florian Angeletti (2022/09/11 13:23 -0700):
> Moving the variable to `Makefile.build_config.in` is definitively a
> good idea.

Cool that you like it!

> Renaming `caml-tex` to `ocamltex` seems fine since it fits with the
> various *tex tools.

Great!

So, I moved on and implemented all this, which also let to a bit of
cleanup in `.gitattributes`.

But now please, Mr Sandman, come!

Octachron commented on Makefile:
```diff
@@ -1507,7 +1507,7 @@ endif
 	  $(MAKE) -C otherlibs/$$i install || exit $$?; \
 	done
-ifneq "$(WITH_OCAMLDOC)" ""
+ifeq "$(build_ocmaldoc)" "true"
```
`build_ocmaldoc` ⇒ `build_ocamldoc`

Octachron commented on Makefile.build_config.in:
```diff
@@ -29,6 +29,20 @@ INSTALL ?= @INSTALL@ -p
+
+build_ocamldoc = @build_ocamldoc@
+
+build_ocamltex=@build_ocamltex@
```
Maybe some missing space here?

Octachron commented:
Beyond the rewriting of the Makefile conditionals, I think that having an explicitly named variables for the status of optional libraries (`lib_$libname`) and optional tools (`build_$toolname`) is an improvement. The new name `ocamltex` also seems more coherent to me.

shindere committed:
Rename caml-tex into ocamltex, + related changes

 * Move the definition of WITH_CAMLTEX from Makefile.config.in to
    Makefile.build_config.in (makes it private)
 * Rename the variable from WITH_CAMLTEX to build_ocamltex
 * Change its definition in configure.ac so that its false rather than
   the empty string when the tool is not built
 * Rename the tool itself from caml-tex to ocamltex

shindere commented:
Florian Angeletti (2022/09/27 05:45 -0700):
> @Octachron commented on this pull request.
>
>
>
> > @@ -1507,7 +1507,7 @@ endif
>  	for i in $(OTHERLIBRARIES); do \
>  	  $(MAKE) -C otherlibs/$$i install || exit $$?; \
>  	done
> -ifneq "$(WITH_OCAMLDOC)" ""
> +ifeq "$(build_ocmaldoc)" "true"
>
> `build_ocmaldoc` ⇒ `build_ocamldoc`

Addressed, thanks!

shindere commented:
Florian Angeletti (2022/09/27 05:57 -0700):
> @Octachron commented on this pull request.
>
>
>
> > @@ -29,6 +29,20 @@ INSTALL ?= @INSTALL@ -p
>  INSTALL_DATA ?= @INSTALL_DATA@
>  INSTALL_PROG ?= @INSTALL_PROGRAM@
>
> +# Whether to build certain libraries and tools
> +
> +build_ocamldebug = @build_ocamldebug@
> +
> +build_ocamldoc = @build_ocamldoc@
> +
> ***@***.***_ocamltex@
>
> Maybe some missing space here?

Yes, indeed. I think I was trying to stick with the convention in
`Makefile.config.in` where we don't use spaces around the `=` sign
because we want to make it easy for pepole to grep the installed
`Make.config` for patterns like `FOO=`.


>
This has been fixed and the branch has been rebased on trunk to resolve
a conflict in a test.

shindere commented:
Florian Angeletti (2022/09/27 06:06 -0700):
> Beyond the rewriting of the Makefile conditionals, I think that having
> an explicitly named variables for the status of optional libraries
> (`lib_$libname`) and optional tools (`build_$toolname`) is an
> improvement. The new name `ocamltex` also seems more coherent to me.

Many thanks for your support! I think once this will be merged we will
be able to slowly stop using and eventually defining a few of our build
variables that are not so nice, e.g. `WITH_OCAMLDEBUG` and the like.

Hopefully this should now be ready to go.

Octachron reviewed (approved):
The current state looks fine to me.

shindere commented:
Florian Angeletti (2022/09/27 06:51 -0700):
> @Octachron approved this pull request.
>
> The current state looks fine to me.

Cool! Thanks!
